File: packages/gpu-view/training/gpu_view/match.py
"""Tolerant surface-form matching of query tokens against a schema.

Mirrors src/match.ts exactly. Both the featurizer and the compiler use it: the
featurizer to say "this token matched a field of kind X", the compiler to turn a
FIELD span back into a real field name. If the two disagree the model is scored
against a field the compiler never picks, so there is exactly one implementation
per language and a shared fixture test.

Matching is over n-grams (1..3 words) of *word* tokens (letter or digit runs).
Hyphens and underscores may sit between the words of a span. Quality tiers, best
first: exact, stem (plural/inflection), prefix (>= 4 chars, unique), typo (>= 5
chars, one edit, unique). Multi-word spans only use exact and stem.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from gpu_utils_training.features import CLASS_DIGIT, CLASS_LETTER, Token, tokenize

EXACT, STEM, INFLECT, PREFIX, TYPO = 0, 1, 2, 3, 4
MIN_PREFIX = 4
MIN_TYPO = 5
MAX_WORDS = 3
CONNECTORS = ("-", "_")

_CAMEL = re.compile(r"([a-z0-9])([A-Z])")


def words_of(surface: str) -> list[str]:
    """Lower-cased letter/digit runs of a surface form; camelCase and snake_case split."""
    text = _CAMEL.sub(r"\1 \2", surface)
    return [t.text.lower() for t in tokenize(text) if t.cls in (CLASS_LETTER, CLASS_DIGIT)]


def stem(word: str) -> str:
    n = len(word)
    if n >= 5 and word.endswith("ies"):
        return word[:-3] + "y"
    if n >= 5 and word.endswith("sses"):
        return word[:-2]
    if n >= 5 and word.endswith("es") and word[-3] in "sxz":
        return word[:-2]
    if n >= 6 and (word.endswith("shes") or word.endswith("ches")):
        return word[:-2]
    if n >= 4 and word.endswith("s") and not word.endswith("ss"):
        return word[:-1]
    return word


def forms(word: str) -> set[str]:
    """Stemmed base forms reachable by stripping comparative/superlative/participle endings.

    "rated" and "rating" share "rat"/"rate"; "cheapest" reaches "cheap"; "downloaded" reaches
    "download". Used by the INFLECT tier so an app can list "cheap" as an alias of price and
    have "cheapest" resolve to it.
    """
    out = {stem(word)}
    n = len(word)
    cands: list[str] = []
    if n >= 6 and word.endswith("iest"):
        cands.append(word[:-4] + "y")
    if n >= 6 and word.endswith("est"):
        cands.append(word[:-3])
    if n >= 5 and word.endswith("ier"):
        cands.append(word[:-3] + "y")
    if n >= 5 and word.endswith("er"):
        cands.append(word[:-2])
    if n >= 5 and word.endswith("ied"):
        cands.append(word[:-3] + "y")
    if n >= 5 and word.endswith("ed"):
        cands.append(word[:-2])
        cands.append(word[:-1])
    if n >= 6 and word.endswith("ing"):
        cands.append(word[:-3])
        cands.append(word[:-3] + "e")
    for c in list(cands):
        if len(c) >= 3 and c[-1] == c[-2] and c[-1] not in "aeiou":
            cands.append(c[:-1])  # bigg → big
    for c in cands:
        if len(c) >= 3:
            out.add(stem(c))
    return out


def within_one_edit(a: str, b: str) -> bool:
    if a == b or abs(len(a) - len(b)) > 1:
        return False
    if len(a) == len(b):
        return sum(1 for x, y in zip(a, b) if x != y) == 1
    short, long = (a, b) if len(a) < len(b) else (b, a)
    i = 0
    while i < len(short) and short[i] == long[i]:
        i += 1
    return short[i:] == long[i + 1 :]


@dataclass(frozen=True)
class Entry:
    words: tuple[str, ...]
    field: int  # index into schema.fields
    kind: str
    alias: bool  # False for the primary name
    value: int = -1  # enum value index (enum entries only)


@dataclass(frozen=True)
class Span:
    start: int  # model-token index, inclusive
    end: int  # model-token index, exclusive
    field: int
    kind: str
    quality: int
    alias: bool
    value: int = -1
    owners: tuple[int, ...] = ()  # enum entries: every field owning this value


def field_entries(schema: dict) -> list[Entry]:
    out: list[Entry] = []
    for i, f in enumerate(schema["fields"]):
        forms = [(f["name"], False)] + [(a, True) for a in f.get("aliases", [])]
        for surface, alias in forms:
            w = tuple(words_of(surface))
            if w and len(w) <= MAX_WORDS:
                out.append(Entry(w, i, f["kind"], alias))
    return out


def enum_entries(schema: dict) -> list[Entry]:
    out: list[Entry] = []
    for i, f in enumerate(schema["fields"]):
        for j, v in enumerate(f.get("values", [])):
            w = tuple(words_of(v))
            if w and len(w) <= MAX_WORDS:
                out.append(Entry(w, i, f["kind"], False, j))
    return out


def model_tokens(text: str) -> list[Token]:
    """Every token except whitespace/newline runs: the sequence the model sees."""
    return [t for t in tokenize(text) if t.cls not in (2, 3)]


def word_positions(tokens: list[Token]) -> list[int]:
    return [i for i, t in enumerate(tokens) if t.cls in (CLASS_LETTER, CLASS_DIGIT)]


def _connected(tokens: list[Token], a: int, b: int) -> bool:
    """True when every token strictly between indices a and b is a connector."""
    for k in range(a + 1, b):
        if tokens[k].text not in CONNECTORS:
            return False
    return True


def _match_at(
    tokens: list[Token], positions: list[int], wi: int, entries: list[Entry], enum: bool
) -> Span | None:
    lowered = [t.text.lower() for t in tokens]
    for n in range(MAX_WORDS, 0, -1):
        if wi + n > len(positions):
            continue
        idx = positions[wi : wi + n]
        ok = all(_connected(tokens, idx[k], idx[k + 1]) for k in range(n - 1))
        if not ok:
            continue
        span_words = tuple(lowered[i] for i in idx)
        stems = tuple(stem(w) for w in span_words)
        start, end = idx[0], idx[-1] + 1
        for quality in (EXACT, STEM):
            hits = [
                e
                for e in entries
                if len(e.words) == n
                and (e.words == span_words if quality == EXACT else tuple(stem(w) for w in e.words) == stems)
            ]
            if hits:
                first = hits[0]
                owners = tuple(sorted({e.field for e in hits}))
                return Span(start, end, first.field, first.kind, quality, first.alias, first.value, owners)
        if n == 1:
            w = span_words[0]
            if len(w) >= 5:
                qforms = forms(w)
                hits = [e for e in entries if len(e.words) == 1 and qforms & forms(e.words[0])]
                if hits:
                    e = hits[0]
                    owners = tuple(sorted({h.field for h in hits}))
                    return Span(start, end, e.field, e.kind, INFLECT, e.alias, e.value, owners)
            if len(w) >= MIN_PREFIX:
                hits = [e for e in entries if len(e.words) == 1 and len(e.words[0]) >= MIN_PREFIX and e.words[0].startswith(w)]
                keys = {(e.field, e.value) for e in hits}
                if len(keys) == 1:
                    e = hits[0]
                    return Span(start, end, e.field, e.kind, PREFIX, e.alias, e.value, (e.field,))
            if len(w) >= MIN_TYPO:
                hits = [e for e in entries if len(e.words) == 1 and within_one_edit(w, e.words[0])]
                keys = {(e.field, e.value) for e in hits}
                if len(keys) == 1:
                    e = hits[0]
                    return Span(start, end, e.field, e.kind, TYPO, e.alias, e.value, (e.field,))
    return None


def match_spans(tokens: list[Token], entries: list[Entry], enum: bool = False) -> list[Span]:
    """Greedy left-to-right, longest-first matching. Spans never overlap."""
    positions = word_positions(tokens)
    spans: list[Span] = []
    wi = 0
    while wi < len(positions):
        found = _match_at(tokens, positions, wi, entries, enum)
        if found is None:
            wi += 1
            continue
        spans.append(found)
        while wi < len(positions) and positions[wi] < found.end:
            wi += 1
    return spans


def resolve_words(words: list[str], entries: list[Entry]) -> Span | None:
    """Resolve an already-extracted phrase (compiler side). Returns a span with fake indices."""
    text = " ".join(words)
    toks = model_tokens(text)
    positions = word_positions(toks)
    if not positions:
        return None
    return _match_at(toks, positions, 0, entries, False)

File: packages/gpu-view/src/match.ts
/**
 * Tolerant surface-form matching of query tokens against a schema.
 * Mirrors training/gpu_view/match.py exactly (fixtures in test/fixtures/match.json).
 *
 * Matches n-grams (1..3 words) of letter/digit tokens, allowing "-" and "_" between
 * words. Quality tiers, best first: exact, stem (plural/inflection), prefix (>= 4
 * chars, unique), typo (>= 5 chars, one edit, unique). Multi-word spans use exact and
 * stem only. Both the featurizer and the compiler call this, so they always agree.
 */
import { CharClass, type Token, tokenize } from "@gpu-utils/runtime";

export type FieldKind = "text" | "number" | "date" | "enum" | "boolean";

export interface SchemaField {
	name: string;
	kind: FieldKind;
	aliases?: string[];
	values?: string[];
}

export interface Schema {
	fields: SchemaField[];
}

export const EXACT = 0;
export const STEM = 1;
export const PREFIX = 2;
export const TYPO = 3;
const MIN_PREFIX = 4;
const MIN_TYPO = 5;
const MAX_WORDS = 3;
const CONNECTORS = new Set(["-", "_"]);

export interface Entry {
	words: string[];
	field: number;
	kind: FieldKind;
	alias: boolean;
	value: number;
}

export interface Span {
	start: number;
	end: number;
	field: number;
	kind: FieldKind;
	quality: number;
	alias: boolean;
	value: number;
	owners: number[];
}

const isWord = (t: Token) =>
	t.cls === CharClass.Letter || t.cls === CharClass.Digit;

/** Lower-cased letter/digit runs of a surface form; camelCase and snake_case split. */
export function wordsOf(surface: string): string[] {
	const text = surface.replace(/([a-z0-9])([A-Z])/g, "$1 $2");
	return tokenize(text)
		.filter(isWord)
		.map((t) => t.text.toLowerCase());
}

export function stem(word: string): string {
	const n = word.length;
	if (n >= 5 && word.endsWith("ies")) return `${word.slice(0, -3)}y`;
	if (n >= 5 && word.endsWith("sses")) return word.slice(0, -2);
	if (n >= 5 && word.endsWith("es") && "sxz".includes(word[n - 3]!))
		return word.slice(0, -2);
	if (n >= 6 && (word.endsWith("shes") || word.endsWith("ches")))
		return word.slice(0, -2);
	if (n >= 4 && word.endsWith("s") && !word.endsWith("ss"))
		return word.slice(0, -1);
	return word;
}

export function withinOneEdit(a: string, b: string): boolean {
	if (a === b || Math.abs(a.length - b.length) > 1) return false;
	if (a.length === b.length) {
		let diff = 0;
		for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) diff++;
		return diff === 1;
	}
	const [short, long] = a.length < b.length ? [a, b] : [b, a];
	let i = 0;
	while (i < short.length && short[i] === long[i]) i++;
	return short.slice(i) === long.slice(i + 1);
}

export function fieldEntries(schema: Schema): Entry[] {
	const out: Entry[] = [];
	schema.fields.forEach((f, i) => {
		const forms: [string, boolean][] = [
			[f.name, false],
			...(f.aliases ?? []).map((a): [string, boolean] => [a, true]),
		];
		for (const [surface, alias] of forms) {
			const words = wordsOf(surface);
			if (words.length > 0 && words.length <= MAX_WORDS)
				out.push({ words, field: i, kind: f.kind, alias, value: -1 });
		}
	});
	return out;
}

export function enumEntries(schema: Schema): Entry[] {
	const out: Entry[] = [];
	schema.fields.forEach((f, i) => {
		(f.values ?? []).forEach((v, j) => {
			const words = wordsOf(v);
			if (words.length > 0 && words.length <= MAX_WORDS)
				out.push({ words, field: i, kind: f.kind, alias: false, value: j });
		});
	});
	return out;
}

/** Every token except whitespace/newline runs: the sequence the model sees. */
export function modelTokens(text: string): Token[] {
	return tokenize(text).filter(
		(t) => t.cls !== CharClass.Space && t.cls !== CharClass.Newline,
	);
}

export function wordPositions(tokens: Token[]): number[] {
	const out: number[] = [];
	tokens.forEach((t, i) => {
		if (isWord(t)) out.push(i);
	});
	return out;
}

function connected(tokens: Token[], a: number, b: number): boolean {
	for (let k = a + 1; k < b; k++)
		if (!CONNECTORS.has(tokens[k]!.text)) return false;
	return true;
}

const same = (a: string[], b: string[]) =>
	a.length === b.length && a.every((w, i) => w === b[i]);

function matchAt(
	tokens: Token[],
	positions: number[],
	wi: number,
	entries: Entry[],
): Span | null {
	const lowered = tokens.map((t) => t.text.toLowerCase());
	for (let n = MAX_WORDS; n >= 1; n--) {
		if (wi + n > positions.length) continue;
		const idx = positions.slice(wi, wi + n);
		let ok = true;
		for (let k = 0; k < n - 1; k++)
			if (!connected(tokens, idx[k]!, idx[k + 1]!)) ok = false;
		if (!ok) continue;
		const spanWords = idx.map((i) => lowered[i]!);
		const stems = spanWords.map(stem);
		const start = idx[0]!;
		const end = idx[n - 1]! + 1;
		for (const quality of [EXACT, STEM]) {
			const hits = entries.filter(
				(e) =>
					e.words.length === n &&
					(quality === EXACT
						? same(e.words, spanWords)
						: same(e.words.map(stem), stems)),
			);
			if (hits.length > 0) {
				const first = hits[0]!;
				const owners = [...new Set(hits.map((e) => e.field))].sort(
					(a, b) => a - b,
				);
				return {
					start,
					end,
					field: first.field,
					kind: first.kind,
					quality,
					alias: first.alias,
					value: first.value,
					owners,
				};
			}
		}
		if (n === 1) {
			const w = spanWords[0]!;
			if (w.length >= MIN_PREFIX) {
				const hits = entries.filter(
					(e) =>
						e.words.length === 1 &&
						e.words[0]!.length >= MIN_PREFIX &&
						e.words[0]!.startsWith(w),
				);
				const keys = new Set(hits.map((e) => `${e.field}:${e.value}`));
				if (keys.size === 1) {
					const e = hits[0]!;
					return {
						start,
						end,
						field: e.field,
						kind: e.kind,
						quality: PREFIX,
						alias: e.alias,
						value: e.value,
						owners: [e.field],
					};
				}
			}
			if (w.length >= MIN_TYPO) {
				const hits = entries.filter(
					(e) => e.words.length === 1 && withinOneEdit(w, e.words[0]!),
				);
				const keys = new Set(hits.map((e) => `${e.field}:${e.value}`));
				if (keys.size === 1) {
					const e = hits[0]!;
					return {
						start,
						end,
						field: e.field,
						kind: e.kind,
						quality: TYPO,
						alias: e.alias,
						value: e.value,
						owners: [e.field],
					};
				}
			}
		}
	}
	return null;
}

/** Greedy left-to-right, longest-first matching. Spans never overlap. */
export function matchSpans(tokens: Token[], entries: Entry[]): Span[] {
	const positions = wordPositions(tokens);
	const spans: Span[] = [];
	let wi = 0;
	while (wi < positions.length) {
		const found = matchAt(tokens, positions, wi, entries);
		if (!found) {
			wi++;
			continue;
		}
		spans.push(found);
		while (wi < positions.length && positions[wi]! < found.end) wi++;
	}
	return spans;
}

/** Resolve an already-extracted phrase (compiler side). Indices are relative to the phrase. */
export function resolveWords(words: string[], entries: Entry[]): Span | null {
	const toks = modelTokens(words.join(" "));
	const positions = wordPositions(toks);
	if (positions.length === 0) return null;
	return matchAt(toks, positions, 0, entries);
}
